ipld/kamt/src/bitfield.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021-2023 Protocol Labs
+// Copyright 2019-2022 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::u64;
+
+use byteorder::{BigEndian, ByteOrder};
+use fvm_ipld_encoding::de::{Deserialize, Deserializer};
+use fvm_ipld_encoding::ser::{Serialize, Serializer};
+use fvm_ipld_encoding::strict_bytes;
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub struct Bitfield([u64; 4]);
+
+impl Serialize for Bitfield {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut v = [0u8; 4 * 8];
+        // Big endian ordering, to match go
+        BigEndian::write_u64(&mut v[..8], self.0[3]);
+        BigEndian::write_u64(&mut v[8..16], self.0[2]);
+        BigEndian::write_u64(&mut v[16..24], self.0[1]);
+        BigEndian::write_u64(&mut v[24..], self.0[0]);
+
+        for i in 0..v.len() {
+            if v[i] != 0 {
+                return strict_bytes::Serialize::serialize(&v[i..], serializer);
+            }
+        }
+
+        <[u8] as strict_bytes::Serialize>::serialize(&[], serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Bitfield {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut res = Bitfield::zero();
+        let bytes = strict_bytes::ByteBuf::deserialize(deserializer)?.into_vec();
+
+        let mut arr = [0u8; 4 * 8];
+        let len = bytes.len();
+        for (old, new) in bytes.iter().zip(arr[(32 - len)..].iter_mut()) {
+            *new = *old;
+        }
+        res.0[3] = BigEndian::read_u64(&arr[..8]);
+        res.0[2] = BigEndian::read_u64(&arr[8..16]);
+        res.0[1] = BigEndian::read_u64(&arr[16..24]);
+        res.0[0] = BigEndian::read_u64(&arr[24..]);
+
+        Ok(res)
+    }
+}
+
+impl Default for Bitfield {
+    fn default() -> Self {
+        Bitfield::zero()
+    }
+}
+
+impl Bitfield {
+    pub fn clear_bit(&mut self, idx: u32) {
+        let ai = idx / 64;
+        let bi = idx % 64;
+        self.0[ai as usize] &= u64::MAX - (1 << bi as u32);
+    }
+
+    pub fn test_bit(&self, idx: u32) -> bool {
+        let ai = idx / 64;
+        let bi = idx % 64;
+
+        self.0[ai as usize] & (1 << bi as u32) != 0
+    }
+
+    pub fn set_bit(&mut self, idx: u32) {
+        let ai = idx / 64;
+        let bi = idx % 64;
+
+        self.0[ai as usize] |= 1 << bi as u32;
+    }
+
+    pub fn count_ones(&self) -> usize {
+        self.0.iter().map(|a| a.count_ones() as usize).sum()
+    }
+
+    pub fn last_one_idx(&self) -> Option<usize> {
+        [192, 128, 64, 0]
+            .iter()
+            .enumerate()
+            .find_map(|(i, offset)| match self.0[3 - i] {
+                0 => None,
+                n => Some((n as f64).log2().floor() as usize + offset),
```

Comment:
You are right, this is ugly. I just wanted to avoid having to loop through all of the indexes to find a non-zero value, I was hoping the logarithm might be faster. I didn't benchmark it, though. 

It doesn't need to be precise since I'm just looking at the highest non-zero bit, and throw away the parts where the conversion from `u64` to `f64` loses precision.

Do let me know if you have a better idea :pray: